### ISSUE TITLE
Autocomplete selecting wrong item

### ISSUE THREAD

**Issue by vron** [dnbexperience/eufemia#1118]
Hi,

I am still struggling with details of the autocomplete component (in async mode searching towards a server that can return different results at different times).

1. When typing to search (first time after loading a page - does not seem to reoccure once i have allready selected) i get some options. If i click the first item the call "await pricingStore.setClient(e.data.selected_key" from below is called with the correct selected_key (as is setVal call one the line above). *However* the autocomplete will show the selected_key from the second item in the list, despite the "console.log("rendering", val);" call showing that the last render was with the val from the first entry.

(I am also 90% sure that it sometimes renders with a completely different value - but can't seem to reliably re-create it now)

2. calling emptyData in the onTypeHandler will completely remove the input the user has typed - I though it should only clear the suggestions data

I am rendering it thus:

    export const ClientFind = observer(() => {
        const [val, setVal] = useState('');
        console.log("rendering", val);
        return (
            <Autocomplete
                value={val}
                mode="async"
                on_type={onTypeHandler}
                on_change={action(async (e) => {
                    if (e.data === undefined) {
                        setVal('');
                        return await pricingStore.setClient(undefined);
                    }
                    setVal(e.data.selected_key);
                    await pricingStore.setClient(e.data.selected_key);
                })}
                no_scroll_animation="true"
            />
        )
    });

using a on_type handler thus:

        const onTypeHandler = ({
            value,
            showIndicator,
            hideIndicator,
            updateData,
            debounce,
            emptyData
        }: any) => {
            showIndicator();
            debounce(
                async ({ value }: any) => {
                    let res = await clientsStore.findClients(value);
                    let data = [...res.map(c => ({
                        selected_key: c.name,
                        selected_value: c.name,
                        content: [
                            c.name,
                            <NumberFormat org>{c.orgno}</NumberFormat>,
                        ],
                    }))];
                    //emptyData();
                    updateData(data);
        
                    hideIndicator();
                }
                , value }, 250
            )
        }

**Comment by vron** [comment]
I should add to make it clear - it "mostly works" since the callback on_change is called with the correct item - so i get the correct key to use further in my logic.

It seems to be after the on_change callback is finished that the autocomplete itself changes the displayed value (the correct one flashes past)

**Comment by tujoworker** [comment]
Can You check if this version works?

Just replace the **verison number** of Eufemia in `package.json` with this https://pkg.csb.dev/dnbexperience/eufemia/commit/12cce683/@dnb/eufemia

**Comment by vron** [comment]
![image](https://user-images.githubusercontent.com/160233/141806013-c5fabb89-95cb-4718-8f10-adf5fb2012b7.png)

using this and running i still see the problem

**Comment by tujoworker** [comment]
I guess you did run `npm / yarn install` after you added this version?

**Comment by vron** [comment]
admittedly no.. I will try again - it's included in the altest release right, so i can just test that?

**Comment by vron** [comment]
Still seeing the bug on .16

**Comment by tujoworker** [comment]
yes, v9.16.0 this fix was included.

**Comment by tujoworker** [comment]
Could you make CSB – we had one with the on_type event before, maybe just make a fork of this one?

**Comment by vron** [comment]
So this is a minimum example to reproduce the bug: https://codesandbox.io/s/eufemia-modal-selection-bug-forked-e86yf?file=/src/App.tsx

1. Load it
2. Type 222 in the search field
3. Click (with mouse) on "Company 1" in the dropdown

Observe: How the "correct" item is selected in the state (renderes the ID of "Company 1" below the search box" but the wrong ID is shown in the autocomplete widget itself ("belonging to company 2")

Without looking at the underlying code it loos like maybe ID's are assigned as numeric values based on some sorting roder - but that different sorts are used in different locations (note that "Company 1" has "222" also in the "shortname" property and not only "name" )

Without looking at th

**Comment by vron** [comment]
forgot to tag you @tujoworker

**Comment by vron** [comment]
Hi - sorry to be a nag - but did you have time to look at the example i created?

**Comment by tujoworker** [comment]
np! I will look at it in the beginning of next week 👍

**Comment by tujoworker** [comment]
@vron I finally got time to look at this. But thank you for the Codesandbox! 🙌

I found a little bug in your code. Remember, the Autocomplete `value` prop will have to be an index of your array you provide, even the one from `updateData`.

Here's a [quick edit](https://codesandbox.io/s/eufemia-modal-selection-not-a-bug-o2lg6?file=/src/App.tsx).

Feel free to open this issue if you find this not to be resolving your reported issue.

Also – let me know how, where and if we can improve the docs ❤️
